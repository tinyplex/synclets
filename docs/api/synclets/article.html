<nav><ul><li><a href="/">Synclets</a></li><li><a href="/api/">API</a></li><li><a href="/api/synclets/"><code>synclets</code></a></li></ul></nav><section class="s1" id="/api/synclets/" data-id="s"><h1><code>synclets</code></h1><p>The <code>synclets</code> module.</p><section class="s2"><h2>Since</h2><p>v0.0.0</p></section><section class="s2" id="/api/synclets/interfaces/" data-id="s/I"><h2>Interfaces</h2><ul><li><a href="/api/synclets/interfaces/other/synclet/"><code>Synclet</code></a></li></ul></section><section class="s2" id="/api/synclets/functions/" data-id="s/F"><h2>Functions</h2><ul><li><a href="/api/synclets/functions/other/createsynclet/"><code>createSynclet</code></a></li></ul></section><section class="s2" id="/api/synclets/type-aliases/" data-id="s/TA"><h2>Type Aliases</h2><ul><li><a href="/api/synclets/type-aliases/other/deletedvalue/"><code>DeletedValue</code></a></li><li><a href="/api/synclets/type-aliases/other/timestamp/"><code>Timestamp</code></a></li><li><a href="/api/synclets/type-aliases/other/value/"><code>Value</code></a></li></ul></section></section>