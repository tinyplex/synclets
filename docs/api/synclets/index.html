<!DOCTYPE html><html lang="en"><head><meta name="viewport" content="width=device-width,initial-scale=1"><link rel="preload" as="image" href="/favicon.svg"><script async="" src="https://www.googletagmanager.com/gtag/js?id=G-14BZ56H85M"></script><link rel="dns-prefetch" href="https://www.googletagmanager.com/"><link href="https://www.googletagmanager.com/gtag/js?id=G-14BZ56H85M" rel="preload" as="script"><link rel="preload" as="font" href="/fonts/inter.woff2" type="font/woff2" crossorigin="anonymous"><link rel="preload" as="font" href="/fonts/inconsolata.woff2" type="font/woff2" crossorigin="anonymous"><link rel="preload" as="font" href="/fonts/shantell.woff2" type="font/woff2" crossorigin="anonymous"><title>synclets | Synclets</title><meta name="description" content="The synclets module."><meta property="og:type" content="website"><meta property="og:title" content="synclets | Synclets"><meta property="og:description" content="The synclets module."><meta property="og:url" content="https://synclets.org/api/synclets/"><meta property="og:image" content="https://synclets.org/favicon_pad.png"><meta name="twitter:card" content="summary"><meta name="twitter:title" content="synclets | Synclets"><meta name="twitter:description" content="The synclets module."><meta name="twitter:site" content="@tinybasejs"><meta name="twitter:image" content="https://synclets.org/favicon_pad.png"><link rel="icon" type="image/svg+xml" href="/favicon.svg"><link rel="canonical" href="https://synclets.org/api/synclets/"><link rel="stylesheet" href="/css/index.css"><script src="/js/app.js"></script></head><body><header><a href="/"><img src="/favicon.svg" alt="Synclets logo"><span>Synclets</span></a><nav><ul><li><a href="/guides/">Guides</a></li><li><a href="/api/">API</a></li><li><a href="https://github.com/tinyplex/synclets">GitHub</a></li></ul></nav><span id="dark"></span></header><main><nav><ul><li id="root" class="parent open"><span></span><a href="/">Synclets</a><ul><li id="G" class="parent"><span></span><a href="/guides/">Guides</a></li><li id="api" class="parent open"><span></span><a href="/api/">API</a><ul><li id="e"><span></span><a href="/api/the-essentials/">The Essentials</a></li><li id="s" class="current parent open"><span></span><a href="/api/synclets/"><code>synclets</code></a><ul><li id="s/I" class="parent"><span></span><a href="/api/synclets/interfaces/">Interfaces</a></li><li id="s/F" class="parent"><span></span><a href="/api/synclets/functions/">Functions</a></li><li id="s/TA" class="parent"><span></span><a href="/api/synclets/type-aliases/">Type Aliases</a></li></ul></li><li id="c2" class="parent"><span></span><a href="/api/connector-base/"><code>connector/base</code></a></li><li id="c7" class="parent"><span></span><a href="/api/connector-file/"><code>connector/file</code></a></li><li id="t" class="parent"><span></span><a href="/api/transport-memory/"><code>transport/memory</code></a></li><li id="u" class="parent"><span></span><a href="/api/utils/"><code>utils</code></a></li></ul></li></ul></li></ul></nav><article><nav><ul><li><a href="/">Synclets</a></li><li><a href="/api/">API</a></li><li><a href="/api/synclets/"><code>synclets</code></a></li></ul></nav><section class="s1" id="/api/synclets/" data-id="s"><h1><code>synclets</code></h1><p>The <code>synclets</code> module.</p><section class="s2"><h2>Since</h2><p>v0.0.0</p></section><section class="s2" id="/api/synclets/interfaces/" data-id="s/I"><h2>Interfaces</h2><ul><li><a href="/api/synclets/interfaces/other/connector/"><code>Connector</code></a></li><li><a href="/api/synclets/interfaces/other/synclet/"><code>Synclet</code></a></li><li><a href="/api/synclets/interfaces/other/transport/methods/other/log/"><code>log</code></a> method</li></ul></section><section class="s2" id="/api/synclets/functions/" data-id="s/F"><h2>Functions</h2><ul><li><a href="/api/synclets/functions/other/getpacketfromparts/"><code>getPacketFromParts</code></a></li><li><a href="/api/synclets/functions/other/getpartsfrompacket/"><code>getPartsFromPacket</code></a></li><li><a href="/api/synclets/functions/other/createconnector/"><code>createConnector</code></a></li><li><a href="/api/synclets/functions/other/createsynclet/"><code>createSynclet</code></a></li><li><a href="/api/synclets/functions/other/createtransport/"><code>createTransport</code></a></li></ul></section><section class="s2" id="/api/synclets/type-aliases/" data-id="s/TA"><h2>Type Aliases</h2><ul><li><a href="/api/synclets/type-aliases/other/address/"><code>Address</code></a></li><li><a href="/api/synclets/type-aliases/other/atom/"><code>Atom</code></a></li><li><a href="/api/synclets/type-aliases/other/connectorimplementations/"><code>ConnectorImplementations</code></a></li><li><a href="/api/synclets/type-aliases/other/connectoroptions/"><code>ConnectorOptions</code></a></li><li><a href="/api/synclets/type-aliases/other/context/"><code>Context</code></a></li><li><a href="/api/synclets/type-aliases/other/hash/"><code>Hash</code></a></li><li><a href="/api/synclets/type-aliases/other/loglevel/"><code>LogLevel</code></a></li><li><a href="/api/synclets/type-aliases/other/logger/"><code>Logger</code></a></li><li><a href="/api/synclets/type-aliases/other/node/"><code>Node</code></a></li><li><a href="/api/synclets/type-aliases/other/subnodes/"><code>SubNodes</code></a></li><li><a href="/api/synclets/type-aliases/other/syncletimplementations/"><code>SyncletImplementations</code></a></li><li><a href="/api/synclets/type-aliases/other/syncletoptions/"><code>SyncletOptions</code></a></li><li><a href="/api/synclets/type-aliases/other/timestamp/"><code>Timestamp</code></a></li><li><a href="/api/synclets/type-aliases/other/timestampandatom/"><code>TimestampAndAtom</code></a></li><li><a href="/api/synclets/type-aliases/other/tomb/"><code>Tomb</code></a></li><li><a href="/api/synclets/type-aliases/other/transportimplementations/"><code>TransportImplementations</code></a></li><li><a href="/api/synclets/type-aliases/other/transportoptions/"><code>TransportOptions</code></a></li></ul></section></section></article><aside></aside></main><footer><nav><a id="gh" href="https://github.com/tinyplex/synclets" target="_blank" rel="noreferrer">GitHub</a><a id="bs" href="https://bsky.app/profile/tinybase.bsky.social" target="_blank" rel="noreferrer">Bluesky</a><a id="tw" href="https://x.com/tinybasejs" target="_blank" rel="noreferrer">X / Twitter</a><a id="dc" href="https://discord.com/invite/mGz3mevwP8" target="_blank" rel="noreferrer">Discord</a></nav><nav><a href="/">Synclets <span id="version"></span></a> Â© 2025-</nav></footer><script>window.dataLayer=window.dataLayer||[];function g(){dataLayer.push(arguments);}g('js',new Date());g('config','G-14BZ56H85M');</script></body></html>