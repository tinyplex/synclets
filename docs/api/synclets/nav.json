{"i":"root","n":"Synclets","u":"/","p":1,"o":1,"_":[{"i":"G","n":"Guides","u":"/guides/","p":1},{"i":"api","n":"API","u":"/api/","p":1,"o":1,"_":[{"i":"e","n":"The Essentials","u":"/api/the-essentials/"},{"i":"s","n":"synclets","u":"/api/synclets/","r":1,"c":1,"p":1,"o":1,"_":[{"i":"s/I","n":"Interfaces","u":"/api/synclets/interfaces/","p":1},{"i":"s/F","n":"Functions","u":"/api/synclets/functions/","p":1},{"i":"s/TA","n":"Type Aliases","u":"/api/synclets/type-aliases/","p":1}]},{"i":"u","n":"utils","u":"/api/utils/","r":1,"p":1},{"i":"c","n":"connector/fs","u":"/api/connector-fs/","r":1,"p":1},{"i":"c2","n":"connector/memory","u":"/api/connector-memory/","r":1,"p":1},{"i":"c3","n":"connector/pglite","u":"/api/connector-pglite/","r":1,"p":1},{"i":"t","n":"transport/memory","u":"/api/transport-memory/","r":1,"p":1},{"i":"t2","n":"transport/ws","u":"/api/transport-ws/","r":1,"p":1},{"i":"s5","n":"server/stateless-ws","u":"/api/server-stateless-ws/","r":1,"p":1}]}]}