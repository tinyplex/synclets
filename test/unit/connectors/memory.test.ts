import {Atoms, ConnectorOptions} from 'synclets';
import {
  createMemoryConnector,
  type MemoryConnector,
} from 'synclets/connector/memory';
import {
  expectDifferingConnectors,
  expectEquivalentConnectors,
  getChainedTestConnectors,
  getPooledTestSyncletsAndConnectors,
  pause,
} from '../common.ts';

interface TestMemoryConnector extends MemoryConnector {
  setAtomForTest(value: string): Promise<void>;
  setNearAtomForTest(value: string): Promise<void>;
  setFarAtomForTest(value: string): Promise<void>;
  delAtomForTest(): Promise<void>;
  getDataForTest(): Atoms;
  getMetaForTest(): string;
}

describe.each([[0, []]])(
  '%d-depth',
  (
    atomDepth: number,
    address: string[],
    nearAddress?: string[],
    farAddress?: string[],
  ) => {
    const createTestMemoryConnector = async (
      options?: ConnectorOptions,
    ): Promise<TestMemoryConnector> => {
      const connector = await createMemoryConnector(atomDepth, {}, options);

      return {
        ...connector,

        setAtomForTest: async (value: string) =>
          connector.setAtom(address, value),
        setNearAtomForTest: async (value: string) => {
          if (nearAddress) {
            connector.setAtom(nearAddress, value);
          }
        },
        setFarAtomForTest: async (value: string) => {
          if (farAddress) {
            connector.setAtom(farAddress, value);
          }
        },

        delAtomForTest: async () => connector.delAtom(address),

        getDataForTest: connector.getAtoms,

        getMetaForTest: connector.getJson,
      };
    };

    describe('2-way', () => {
      test('connected, initial', async () => {
        const [[, connector1], [, connector2]] =
          await getPooledTestSyncletsAndConnectors(
            createTestMemoryConnector,
            2,
          );

        expectEquivalentConnectors([connector1, connector2]);
      });

      test('connected', async () => {
        const [[, connector1], [, connector2]] =
          await getPooledTestSyncletsAndConnectors(
            createTestMemoryConnector,
            2,
          );

        await connector1.setAtomForTest('A');
        expectEquivalentConnectors([connector1, connector2]);

        await connector2.setAtomForTest('B');
        expectEquivalentConnectors([connector1, connector2]);
      });

      test('connected, deletion', async () => {
        const [[, connector1], [, connector2]] =
          await getPooledTestSyncletsAndConnectors(
            createTestMemoryConnector,
            2,
          );

        await connector1.setAtomForTest('A');
        expectEquivalentConnectors([connector1, connector2]);

        const timestamp = connector1.getMetaForTest();
        await connector1.delAtomForTest();
        expectEquivalentConnectors([connector1, connector2]);
        expect(timestamp).not.toEqual(connector1.getMetaForTest());
      });

      test('start 1, set 1, start 2', async () => {
        const [[synclet1, connector1], [synclet2, connector2]] =
          await getPooledTestSyncletsAndConnectors(
            createTestMemoryConnector,
            2,
            false,
          );

        await synclet1.start();

        await connector1.setAtomForTest('A');
        expectDifferingConnectors(connector1, connector2);

        await synclet2.start();
        expectEquivalentConnectors([connector1, connector2]);
      });

      test('start 2, set 1, start 1', async () => {
        const [[synclet1, connector1], [synclet2, connector2]] =
          await getPooledTestSyncletsAndConnectors(
            createTestMemoryConnector,
            2,
            false,
          );

        await synclet2.start();
        await connector1.connect();
        await connector1.setAtomForTest('A');
        expectDifferingConnectors(connector1, connector2);

        await synclet1.start();
        expectEquivalentConnectors([connector1, connector2]);
      });

      test('stop 1, set 1, start 1', async () => {
        const [[synclet1, connector1], [synclet2, connector2]] =
          await getPooledTestSyncletsAndConnectors(
            createTestMemoryConnector,
            2,
            false,
          );

        await synclet1.start();
        await synclet2.start();

        await connector1.setAtomForTest('A');
        expectEquivalentConnectors([connector1, connector2]);

        await synclet1.stop();
        await connector1.connect();
        await connector1.setAtomForTest('B');
        expectDifferingConnectors(connector1, connector2);

        await synclet1.start();
        expectEquivalentConnectors([connector1, connector2]);
      });

      test('stop 1, set 2, start 1', async () => {
        const [[synclet1, connector1], [synclet2, connector2]] =
          await getPooledTestSyncletsAndConnectors(
            createTestMemoryConnector,
            2,
            false,
          );

        await synclet1.start();
        await synclet2.start();

        await connector1.setAtomForTest('A');
        expectEquivalentConnectors([connector1, connector2]);

        await synclet1.stop();
        await connector2.setAtomForTest('B');
        expectDifferingConnectors(connector1, connector2);

        await synclet1.start();
        expectEquivalentConnectors([connector1, connector2]);
      });

      test('set 1, set 2, start 2, start 1', async () => {
        const [[synclet1, connector1], [synclet2, connector2]] =
          await getPooledTestSyncletsAndConnectors(
            createTestMemoryConnector,
            2,
            false,
          );

        await connector1.connect();
        await connector1.setAtomForTest('A');
        expectDifferingConnectors(connector1, connector2);

        await pause();

        await connector2.connect();
        await connector2.setAtomForTest('B');
        expectDifferingConnectors(connector1, connector2);

        await synclet2.start();
        await synclet1.start();
        expectEquivalentConnectors([connector1, connector2]);
      });

      test('connected, near atom', async () => {
        if (nearAddress) {
          const [[, connector1], [, connector2]] =
            await getPooledTestSyncletsAndConnectors(
              createTestMemoryConnector,
              2,
            );
          await connector1.setAtomForTest('A');
          await connector2.setNearAtomForTest('B');
          expectEquivalentConnectors([connector1, connector2], {
            t1: {r1: {c1: 'A', c2: 'B'}},
          });
        }
      });

      test('connected, far atom', async () => {
        if (farAddress) {
          const [[, connector1], [, connector2]] =
            await getPooledTestSyncletsAndConnectors(
              createTestMemoryConnector,
              2,
            );
          await connector1.setAtomForTest('A');
          await connector2.setFarAtomForTest('B');
          expectEquivalentConnectors([connector1, connector2], {
            t1: {r1: {c1: 'A'}},
            t2: {r2: {c2: 'B'}},
          });
        }
      });

      test('disconnected, near atom', async () => {
        if (nearAddress) {
          const [[synclet1, connector1], [synclet2, connector2]] =
            await getPooledTestSyncletsAndConnectors(
              createTestMemoryConnector,
              2,
              false,
            );
          await connector1.connect();
          await connector2.connect();
          await connector1.setAtomForTest('A');
          await connector2.setNearAtomForTest('B');
          await synclet1.start();
          await synclet2.start();
          expectEquivalentConnectors([connector1, connector2], {
            t1: {r1: {c1: 'A', c2: 'B'}},
          });
        }
      });

      test('disconnected, far atom', async () => {
        if (farAddress) {
          const [[synclet1, connector1], [synclet2, connector2]] =
            await getPooledTestSyncletsAndConnectors(
              createTestMemoryConnector,
              2,
              false,
            );
          await connector1.connect();
          await connector2.connect();
          await connector1.setAtomForTest('A');
          await connector2.setFarAtomForTest('B');
          await synclet1.start();
          await synclet2.start();
          expectEquivalentConnectors([connector1, connector2], {
            t1: {r1: {c1: 'A'}},
            t2: {r2: {c2: 'B'}},
          });
        }
      });

      test('disconnected, conflicting values', async () => {
        const [[synclet1, connector1], [synclet2, connector2]] =
          await getPooledTestSyncletsAndConnectors(
            createTestMemoryConnector,
            2,
            false,
          );
        await connector1.connect();
        await connector2.connect();
        await connector1.setAtomForTest('A');
        await pause();
        await connector1.setAtomForTest('B');
        await synclet1.start();
        await synclet2.start();
        expectEquivalentConnectors([connector1, connector2], {
          t1: {r1: {c1: 'B'}},
        });
      });

      test('disconnected, conflicting values 2', async () => {
        if (nearAddress && farAddress) {
          const [[synclet1, connector1], [synclet2, connector2]] =
            await getPooledTestSyncletsAndConnectors(
              createTestMemoryConnector,
              2,
              false,
            );
          await connector1.connect();
          await connector2.connect();
          await connector1.setAtomForTest('A');
          await connector1.setNearAtomForTest('B');
          await pause();
          await connector2.setNearAtomForTest('C');
          await connector2.setFarAtomForTest('D');
          await synclet1.start();
          await synclet2.start();
          expectEquivalentConnectors([connector1, connector2], {
            t1: {r1: {c1: 'A', c2: 'C'}},
            t2: {r2: {c2: 'D'}},
          });
        }
      });
    });

    describe.each([3, 10])('%d-way', (count: number) => {
      test('pool', async () => {
        const syncletsAndConnectors = await getPooledTestSyncletsAndConnectors(
          createTestMemoryConnector,
          count,
        );

        const connectors = syncletsAndConnectors.map(
          ([, connector]) => connector,
        );

        for (const [i, connector] of connectors.entries()) {
          await connector.setAtomForTest('A' + (i + 1));
          expectEquivalentConnectors(connectors);
        }
      });

      test('chain', async () => {
        const connectors = await getChainedTestConnectors(
          createTestMemoryConnector,
          count,
        );

        for (const [i, connector] of connectors.entries()) {
          await connector.setAtomForTest('V' + (i + 1));
          expectEquivalentConnectors(connectors);
        }
      });

      test('ring', async () => {
        const connectors = await getChainedTestConnectors(
          createTestMemoryConnector,
          count,
          true,
        );

        for (const [i, connector] of connectors.entries()) {
          await connector.setAtomForTest('V' + (i + 1));
          expectEquivalentConnectors(connectors);
        }
      });
    });
  },
);
