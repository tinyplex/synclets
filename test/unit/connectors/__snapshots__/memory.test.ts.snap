// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`0-depth 2-way connected, deletion: equivalent 1`] = `"A"`;

exports[`0-depth 2-way connected, deletion: equivalent 2`] = `undefined`;

exports[`0-depth 2-way connected, initial: equivalent 1`] = `undefined`;

exports[`0-depth 2-way connected: equivalent 1`] = `"A"`;

exports[`0-depth 2-way connected: equivalent 2`] = `"B"`;

exports[`0-depth 2-way disconnected, conflicting values: equivalent 1`] = `"B"`;

exports[`0-depth 2-way set 1, set 2, start 2, start 1: differing 1`] = `
[
  "A",
  undefined,
]
`;

exports[`0-depth 2-way set 1, set 2, start 2, start 1: differing 2`] = `
[
  "A",
  "B",
]
`;

exports[`0-depth 2-way set 1, set 2, start 2, start 1: equivalent 1`] = `"B"`;

exports[`0-depth 2-way start 1, set 1, start 2: differing 1`] = `
[
  "A",
  undefined,
]
`;

exports[`0-depth 2-way start 1, set 1, start 2: equivalent 1`] = `"A"`;

exports[`0-depth 2-way start 2, set 1, start 1: differing 1`] = `
[
  "A",
  undefined,
]
`;

exports[`0-depth 2-way start 2, set 1, start 1: equivalent 1`] = `"A"`;

exports[`0-depth 2-way stop 1, set 1, start 1: differing 1`] = `
[
  "B",
  "A",
]
`;

exports[`0-depth 2-way stop 1, set 1, start 1: equivalent 1`] = `"A"`;

exports[`0-depth 2-way stop 1, set 1, start 1: equivalent 2`] = `"B"`;

exports[`0-depth 2-way stop 1, set 2, start 1: differing 1`] = `
[
  "A",
  "B",
]
`;

exports[`0-depth 2-way stop 1, set 2, start 1: equivalent 1`] = `"A"`;

exports[`0-depth 2-way stop 1, set 2, start 1: equivalent 2`] = `"B"`;

exports[`0-depth 3-way chain: equivalent 1`] = `"A1"`;

exports[`0-depth 3-way chain: equivalent 2`] = `"A2"`;

exports[`0-depth 3-way chain: equivalent 3`] = `"A3"`;

exports[`0-depth 3-way pool: equivalent 1`] = `"A1"`;

exports[`0-depth 3-way pool: equivalent 2`] = `"A2"`;

exports[`0-depth 3-way pool: equivalent 3`] = `"A3"`;

exports[`0-depth 3-way ring: equivalent 1`] = `"A1"`;

exports[`0-depth 3-way ring: equivalent 2`] = `"A2"`;

exports[`0-depth 3-way ring: equivalent 3`] = `"A3"`;

exports[`0-depth 10-way chain: equivalent 1`] = `"A1"`;

exports[`0-depth 10-way chain: equivalent 2`] = `"A2"`;

exports[`0-depth 10-way chain: equivalent 3`] = `"A3"`;

exports[`0-depth 10-way chain: equivalent 4`] = `"A4"`;

exports[`0-depth 10-way chain: equivalent 5`] = `"A5"`;

exports[`0-depth 10-way chain: equivalent 6`] = `"A6"`;

exports[`0-depth 10-way chain: equivalent 7`] = `"A7"`;

exports[`0-depth 10-way chain: equivalent 8`] = `"A8"`;

exports[`0-depth 10-way chain: equivalent 9`] = `"A9"`;

exports[`0-depth 10-way chain: equivalent 10`] = `"A10"`;

exports[`0-depth 10-way pool: equivalent 1`] = `"A1"`;

exports[`0-depth 10-way pool: equivalent 2`] = `"A2"`;

exports[`0-depth 10-way pool: equivalent 3`] = `"A3"`;

exports[`0-depth 10-way pool: equivalent 4`] = `"A4"`;

exports[`0-depth 10-way pool: equivalent 5`] = `"A5"`;

exports[`0-depth 10-way pool: equivalent 6`] = `"A6"`;

exports[`0-depth 10-way pool: equivalent 7`] = `"A7"`;

exports[`0-depth 10-way pool: equivalent 8`] = `"A8"`;

exports[`0-depth 10-way pool: equivalent 9`] = `"A9"`;

exports[`0-depth 10-way pool: equivalent 10`] = `"A10"`;

exports[`0-depth 10-way ring: equivalent 1`] = `"A1"`;

exports[`0-depth 10-way ring: equivalent 2`] = `"A2"`;

exports[`0-depth 10-way ring: equivalent 3`] = `"A3"`;

exports[`0-depth 10-way ring: equivalent 4`] = `"A4"`;

exports[`0-depth 10-way ring: equivalent 5`] = `"A5"`;

exports[`0-depth 10-way ring: equivalent 6`] = `"A6"`;

exports[`0-depth 10-way ring: equivalent 7`] = `"A7"`;

exports[`0-depth 10-way ring: equivalent 8`] = `"A8"`;

exports[`0-depth 10-way ring: equivalent 9`] = `"A9"`;

exports[`0-depth 10-way ring: equivalent 10`] = `"A10"`;

exports[`1-depth 2-way connected, deletion: equivalent 1`] = `
{
  "a": "A",
}
`;

exports[`1-depth 2-way connected, deletion: equivalent 2`] = `
{
  "a": undefined,
}
`;

exports[`1-depth 2-way connected, initial: equivalent 1`] = `{}`;

exports[`1-depth 2-way connected, near atom: equivalent 1`] = `
{
  "a": "A",
  "b": "B",
}
`;

exports[`1-depth 2-way connected: equivalent 1`] = `
{
  "a": "A",
}
`;

exports[`1-depth 2-way connected: equivalent 2`] = `
{
  "a": "B",
}
`;

exports[`1-depth 2-way disconnected, conflicting values: equivalent 1`] = `
{
  "a": "B",
}
`;

exports[`1-depth 2-way disconnected, near atom: equivalent 1`] = `
{
  "a": "A",
  "b": "B",
}
`;

exports[`1-depth 2-way set 1, set 2, start 2, start 1: differing 1`] = `
[
  {
    "a": "A",
  },
  {},
]
`;

exports[`1-depth 2-way set 1, set 2, start 2, start 1: differing 2`] = `
[
  {
    "a": "A",
  },
  {
    "a": "B",
  },
]
`;

exports[`1-depth 2-way set 1, set 2, start 2, start 1: equivalent 1`] = `
{
  "a": "B",
}
`;

exports[`1-depth 2-way start 1, set 1, start 2: differing 1`] = `
[
  {
    "a": "A",
  },
  {},
]
`;

exports[`1-depth 2-way start 1, set 1, start 2: equivalent 1`] = `
{
  "a": "A",
}
`;

exports[`1-depth 2-way start 2, set 1, start 1: differing 1`] = `
[
  {
    "a": "A",
  },
  {},
]
`;

exports[`1-depth 2-way start 2, set 1, start 1: equivalent 1`] = `
{
  "a": "A",
}
`;

exports[`1-depth 2-way stop 1, set 1, start 1: differing 1`] = `
[
  {
    "a": "B",
  },
  {
    "a": "A",
  },
]
`;

exports[`1-depth 2-way stop 1, set 1, start 1: equivalent 1`] = `
{
  "a": "A",
}
`;

exports[`1-depth 2-way stop 1, set 1, start 1: equivalent 2`] = `
{
  "a": "B",
}
`;

exports[`1-depth 2-way stop 1, set 2, start 1: differing 1`] = `
[
  {
    "a": "A",
  },
  {
    "a": "B",
  },
]
`;

exports[`1-depth 2-way stop 1, set 2, start 1: equivalent 1`] = `
{
  "a": "A",
}
`;

exports[`1-depth 2-way stop 1, set 2, start 1: equivalent 2`] = `
{
  "a": "B",
}
`;

exports[`1-depth 3-way chain: equivalent 1`] = `
{
  "a": "A1",
}
`;

exports[`1-depth 3-way chain: equivalent 2`] = `
{
  "a": "A2",
}
`;

exports[`1-depth 3-way chain: equivalent 3`] = `
{
  "a": "A3",
}
`;

exports[`1-depth 3-way pool: equivalent 1`] = `
{
  "a": "A1",
}
`;

exports[`1-depth 3-way pool: equivalent 2`] = `
{
  "a": "A2",
}
`;

exports[`1-depth 3-way pool: equivalent 3`] = `
{
  "a": "A3",
}
`;

exports[`1-depth 3-way ring: equivalent 1`] = `
{
  "a": "A1",
}
`;

exports[`1-depth 3-way ring: equivalent 2`] = `
{
  "a": "A2",
}
`;

exports[`1-depth 3-way ring: equivalent 3`] = `
{
  "a": "A3",
}
`;

exports[`1-depth 10-way chain: equivalent 1`] = `
{
  "a": "A1",
}
`;

exports[`1-depth 10-way chain: equivalent 2`] = `
{
  "a": "A2",
}
`;

exports[`1-depth 10-way chain: equivalent 3`] = `
{
  "a": "A3",
}
`;

exports[`1-depth 10-way chain: equivalent 4`] = `
{
  "a": "A4",
}
`;

exports[`1-depth 10-way chain: equivalent 5`] = `
{
  "a": "A5",
}
`;

exports[`1-depth 10-way chain: equivalent 6`] = `
{
  "a": "A6",
}
`;

exports[`1-depth 10-way chain: equivalent 7`] = `
{
  "a": "A7",
}
`;

exports[`1-depth 10-way chain: equivalent 8`] = `
{
  "a": "A8",
}
`;

exports[`1-depth 10-way chain: equivalent 9`] = `
{
  "a": "A9",
}
`;

exports[`1-depth 10-way chain: equivalent 10`] = `
{
  "a": "A10",
}
`;

exports[`1-depth 10-way pool: equivalent 1`] = `
{
  "a": "A1",
}
`;

exports[`1-depth 10-way pool: equivalent 2`] = `
{
  "a": "A2",
}
`;

exports[`1-depth 10-way pool: equivalent 3`] = `
{
  "a": "A3",
}
`;

exports[`1-depth 10-way pool: equivalent 4`] = `
{
  "a": "A4",
}
`;

exports[`1-depth 10-way pool: equivalent 5`] = `
{
  "a": "A5",
}
`;

exports[`1-depth 10-way pool: equivalent 6`] = `
{
  "a": "A6",
}
`;

exports[`1-depth 10-way pool: equivalent 7`] = `
{
  "a": "A7",
}
`;

exports[`1-depth 10-way pool: equivalent 8`] = `
{
  "a": "A8",
}
`;

exports[`1-depth 10-way pool: equivalent 9`] = `
{
  "a": "A9",
}
`;

exports[`1-depth 10-way pool: equivalent 10`] = `
{
  "a": "A10",
}
`;

exports[`1-depth 10-way ring: equivalent 1`] = `
{
  "a": "A1",
}
`;

exports[`1-depth 10-way ring: equivalent 2`] = `
{
  "a": "A2",
}
`;

exports[`1-depth 10-way ring: equivalent 3`] = `
{
  "a": "A3",
}
`;

exports[`1-depth 10-way ring: equivalent 4`] = `
{
  "a": "A4",
}
`;

exports[`1-depth 10-way ring: equivalent 5`] = `
{
  "a": "A5",
}
`;

exports[`1-depth 10-way ring: equivalent 6`] = `
{
  "a": "A6",
}
`;

exports[`1-depth 10-way ring: equivalent 7`] = `
{
  "a": "A7",
}
`;

exports[`1-depth 10-way ring: equivalent 8`] = `
{
  "a": "A8",
}
`;

exports[`1-depth 10-way ring: equivalent 9`] = `
{
  "a": "A9",
}
`;

exports[`1-depth 10-way ring: equivalent 10`] = `
{
  "a": "A10",
}
`;
